docs/design/coreclr/botr/data_contracts_for_coreclr.md

```diff
@@ -0,0 +1,430 @@
+# Data Contracts for CoreCLR, and the data contract stream
+
+Data contracts represent the manner in which a tool which is not the runtime can reliably understand and observe the behavior of the runtime. Contracts are defined by their documentation, and the runtime describes what contracts are applicable to understanding that runtime.
+
+## Data Contract Stream
+The physical layout of this stream is not defined in this document, but its practical effects are.
+
+The data contract stream has a set of records of the following forms.
+
+### Global Values
+Global values which can be of types (int8, uint8, int16, uint16, int32, uint32, int64, uint64, pointer, unicode string)
+All global values have a unicode string describing their name, and a value of one of the above types.
+
+### Compatible Contract
+Each compatible contract is described by a unicode string naming the contract, and a uint32 version. It is an ERROR if multiple versions of a contract are specified in the stream.
```

Comment:
I'm not convinced we need human readable names in the binary runtime artifact. (In the source code and in the docs, sure).  I think we should just have a blob of GUIDs where each "name,version" is a new GUID.

Reply:
> I'm not convinced we need human readable names in the binary runtime artifact.

I think I disagree. The binary artifact should be self-contained and as such have human readable mappings. This can be optimized in some ways. See the Fall prototype for how an ordinal to string mapping was created and then ordinals used in all other places. Being able to intelligently debug this code is going to be paramount. I could see an argument for having an debugger extension with this mapping, but I would like to see the mechanics of that. It is something to explore.

Reply:
> I think I disagree. The binary artifact should be self-contained and as such have human readable mappings. This can be optimized in some ways. See the Fall prototype for how an ordinal to string mapping was created and then ordinals used in all other places. Being able to intelligently debug this code is going to be paramount.

If you're debugging the data contract reader, you have a mapping from GUIDs to names in the reader.  It might make sense for the reader to immediately convert the GUIDs to human-readable names as soon as it connects to a target process.   That doesn't mean the physical representation should serialize human-readable names.

> I could see an argument for having an debugger extension with this mapping, but I would like to see the mechanics of that. It is something to explore.

you wouldn't need any extension - by the time you're debugging the reader, you can work in terms of names.

Reply:
> If you're debugging the data contract reader, you have a mapping from GUIDs to names in the reader.

The reader then becomes imbued with the totality of history instead of just working on the provided data. Note how the Fall prototype consumed all the human readable details from the current runtime and could ignore anything it didn't know about.

Reply:
> The reader then becomes imbued with the totality of history instead of just working on the provided data. 

yes, of course.  My understanding was that the data contract reader proposal was that the reader is capable of working with any version going back to the first release of the data contract work.

> Note how the Fall prototype consumed all the human readable details from the current runtime and could ignore anything it didn't know about.

I think that's fine for data but not for algorithms.

Reply:
I have preference for human readable names as well. They make things easy to debug and diagnose. The GUIDs are not that much more compact compared to human readable names.

If we push it to the limit, we can even use JSON to represent most of the data contract descriptor. (The JSON would be embedded as a read-only data blob in the executable.)

Reply:
As long as we agree that we don't expect that many contracts then I don't think we need to be overly efficient with the encoding. Two potential encodings that I believe would be easy to work with that aren't that large:

1. Define an ordinal for each contract and put the name into the header structure definition as the field name. IMO this is the simplest approach that doesn't have much overhead unless we start creating and deprecating large numbers of contracts:

```
struct DotNetDebugHeader
{
    // cookie and flags fields omitted
    int CountOfContracts; // Increase this number when we append new contracts
    void* MethodTableContract;
    void* GCContract;
    void* ThreadContract;
    ...
}
```

2. If we expect it is frequent to deprecate entire contracts then make an array of them and assign each one a modest fixed size name.

```
struct DotNetDebugHeader
{
    // cookie and flags fields omitted
    Contract* Contracts;
    int CountOfContracts;
    ...
}

struct Contract
{
    char[12] Name;  //UTF8 null terminated friendly name
                    //Either we pick short names for contracts or we associate the occasional 
                    //long name with a shorter abbreviation.
    int Version;
    // whatever other per-contract data goes here
}
```